### PR TITLE
Add Dependencies for Arrow's Azure Filesystem

### LINKED ISSUE #1430: Conda Arrow Missing Azure FileSystem Support
### Comment:

Arrow 15 and 16 added support for Azure's various filesystems using the Azure C++ package (as taken from their `vcpkg.json`):
```
    "azure": {
      "description": "Azure blob storage support",
      "dependencies": [
        "azure-core-cpp",
        "azure-identity-cpp",
        "azure-storage-blobs-cpp",
        "azure-storage-common-cpp",
        "azure-storage-files-datalake-cpp"
      ]
    },
```

Conda's version of Arrow does not depend on these libraries, so it is not built with Azure support. Similarly, PyArrow lacks Azure support as well, with error:
```
ImportError: The pyarrow installation is not built with support for 'AzureFileSystem'
```

As far as I can tell, adding these libraries to the recipe will fix this issue.

### PR DESCRIPTION
**Pull request by srilman**
Resolved #1430

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by h-vetinari** [comment]
Thanks for the PR! You can try `-DCMAKE_UNITY_BUILD=OFF ^` in `bld.bat`, that should at least give better error messages.

**Comment by srilman** [comment]
Main Error Message:
```
error C2027: use of undefined type 'arrow::fs::AzureFileSystem::Impl'
```

Here is the full error message:

```
FAILED: src/arrow/CMakeFiles/arrow_filesystem_shared.dir/filesystem/filesystem.cc.obj 
C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\cl.exe  /nologo /TP -DABSL_CONSUME_DLL -DARROW_EXPORTING -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_WITH_TIMING_TESTS -DAWS_AUTH_USE_IMPORT_EXPORT -DAWS_CAL_USE_IMPORT_EXPORT -DAWS_CHECKSUMS_USE_IMPORT_EXPORT -DAWS_COMMON_USE_IMPORT_EXPORT -DAWS_COMPRESSION_USE_IMPORT_EXPORT -DAWS_CRT_CPP_USE_IMPORT_EXPORT -DAWS_EVENT_STREAM_USE_IMPORT_EXPORT -DAWS_HTTP_USE_IMPORT_EXPORT -DAWS_IO_USE_IMPORT_EXPORT -DAWS_MQTT_USE_IMPORT_EXPORT -DAWS_S3_USE_IMPORT_EXPORT -DAWS_SDKUTILS_USE_IMPORT_EXPORT -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=11 -DAWS_SDK_VERSION_PATCH=329 -DAWS_USE_IO_COMPLETION_PORTS -DAZ_CORE_DLL -DAZ_IDENTITY_DLL -DAZ_RTTI -DAZ_STORAGE_BLOBS_DLL -DAZ_STORAGE_COMMON_DLL -DAZ_STORAGE_FILES_DATALAKE_DLL -DPROTOBUF_USE_DLLS -DUSE_IMPORT_EXPORT -DUSE_IMPORT_EXPORT=1 -DUSE_WINDOWS_DLL_SEMANTICS -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -I%SRC_DIR%\cpp\build\src -I%SRC_DIR%\cpp\src -I%SRC_DIR%\cpp\src\generated -external:I%PREFIX%\Library\include -external:I%SRC_DIR%\cpp\thirdparty\hadoop\include -external:W0 /DWIN32 /D_WINDOWS /GR /EHsc /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING  /EHsc /wd5105 /bigobj /utf-8 /W3 /wd4800 /wd4996 /wd4065  /O2 /Ob2 /DNDEBUG -std:c++17 -MD /showIncludes /Fosrc\arrow\CMakeFiles\arrow_filesystem_shared.dir\filesystem\filesystem.cc.obj /Fdsrc\arrow\CMakeFiles\arrow_filesystem_shared.dir\ /FS -c %SRC_DIR%\cpp\src\arrow\filesystem\filesystem.cc
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3089): error C2027: use of undefined type 'arrow::fs::AzureFileSystem::Impl'
%SRC_DIR%\cpp\src\arrow/filesystem/azurefs.h(232): note: see declaration of 'arrow::fs::AzureFileSystem::Impl'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3088): note: while compiling class template member function 'void std::default_delete<arrow::fs::AzureFileSystem::Impl>::operator ()(_Ty *) noexcept const'
        with
        [
            _Ty=arrow::fs::AzureFileSystem::Impl
        ]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3198): note: see reference to function template instantiation 'void std::default_delete<arrow::fs::AzureFileSystem::Impl>::operator ()(_Ty *) noexcept const' being compiled
        with
        [
            _Ty=arrow::fs::AzureFileSystem::Impl
        ]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3125): note: see reference to class template instantiation 'std::default_delete<arrow::fs::AzureFileSystem::Impl>' being compiled
%SRC_DIR%\cpp\src\arrow/filesystem/azurefs.h(233): note: see reference to class template instantiation 'std::unique_ptr<arrow::fs::AzureFileSystem::Impl,std::default_delete<arrow::fs::AzureFileSystem::Impl>>' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3089): error C2338: can't delete an incomplete type
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3090): warning C4150: deletion of pointer to incomplete type 'arrow::fs::AzureFileSystem::Impl'; no destructor called
%SRC_DIR%\cpp\src\arrow/filesystem/azurefs.h(232): note: see declaration of 'arrow::fs::AzureFileSystem::Impl'
```

File: recipe/bld.bat [hunk 1]
```diff
@@ -17,6 +17,7 @@ set "READ_RECIPE_META_YAML_WHY_NOT=OFF"
 :: https://github.com/apache/arrow/blame/apache-arrow-12.0.0/cpp/cmake_modules/DefineOptions.cmake
 cmake -G "Ninja" ^
       -DARROW_ACERO=ON ^
+      -DARROW_AZURE=ON ^
```
Comment (h-vetinari) [inline recipe/bld.bat:20]: You _could_ set this to `OFF`

File: recipe/meta.yaml [hunk 2]
```diff
@@ -58,6 +58,12 @@ requirements:
     - llvmdev {{ llvm_version }}
     - aws-crt-cpp
     - aws-sdk-cpp
+    # azure filesystem dependencies
+    - azure-core-cpp
+    - azure-identity-cpp
+    - azure-storage-blobs-cpp
+    - azure-storage-common-cpp
+    - azure-storage-files-datalake-cpp
```
Comment (h-vetinari) [inline recipe/meta.yaml:66]: and restrict the dependencies here
```suggestion
    - azure-core-cpp                    # [unix]
    - azure-identity-cpp                # [unix]
    - azure-storage-blobs-cpp           # [unix]
    - azure-storage-common-cpp          # [unix]
    - azure-storage-files-datalake-cpp  # [unix]
```

**Review by h-vetinari — COMMENTED** [review]
It's possible to exclude windows, but unless the urgency here is unusually high, I'd like to figure out why the windows builds are failing, and fix it. Could you raise an upstream issue with the error message here? It's possible to backport patches if/once they appear.

**Comment by srilman** [comment]
For my use case, it is urgent yes, so I would prefer to do so if possible. I will make the changes you suggested as well as post this issue in the Arrow repo.

File: .ci_support/linux_64_c_compiler_version11c_stdlib_version2.17cuda_compilernvcccuda_compiler_version11.8cxx_compiler_version11.yaml [hunk 3]
```diff
@@ -2,6 +2,16 @@ aws_crt_cpp:
...
+azure_core_cpp:
+- 1.11.1
+azure_identity_cpp:
+- 1.6.0
+azure_storage_blobs_cpp:
+- 12.10.0
+azure_storage_common_cpp:
+- 12.5.0
+azure_storage_files_datalake_cpp:
+- 12.9.0
```
Comment (h-vetinari) [inline .ci_support/linux_64_c_compiler_version11c_stdlib_version2.17cuda_compilernvcccuda_compiler_version11.8cxx_compiler_version11.yaml:14]: You cannot set this manually, it needs to be done correctly: https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6003

**Review by h-vetinari — CHANGES_REQUESTED** [review]
I've started picking your changes into #1435, but we're not ready to merge this until someone adds the azure-deps to the global pinning, and cleans up the azure-* feedstocks so they're all built for a consistent set of versions.

**Comment by srilman** [comment]
@h-vetinari To be honest, I don't really understand what that means. Is that something I could help with?

**Comment by h-vetinari** [comment]
> @h-vetinari To be honest, I don't really understand what you mean by global pinning.

It is what's under https://github.com/conda-forge/conda-forge-pinning-feedstock, specifically [here](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml). 

> Is that something I could help with? Is there anything I can do about the azure dep versionings as well? As I said, would like to get this in ASAP

You can help the maintainers of the `azure-*` feedstocks clean up the state of their feedstocks by helping with https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6003.

**Comment by h-vetinari** [comment]
Tried the migrator from https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6056, but it's not ready yet (this is exactly why we need https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6003)
```
Could not solve for environment specs
The following packages are incompatible
├─ azure-core-cpp 1.12.0.*  is requested and can be installed;
├─ azure-identity-cpp 1.8.0.*  is installable and it requires
│  └─ azure-core-cpp >=1.12.0,<1.12.1.0a0 , which can be installed;
├─ azure-storage-blobs-cpp 12.11.0.*  is not installable because it requires
│  └─ azure-core-cpp >=1.11.3,<1.11.4.0a0 , which conflicts with any installable versions previously reported;
└─ azure-storage-files-datalake-cpp 12.10.0.*  is not installable because it requires
   └─ azure-core-cpp >=1.11.3,<1.11.4.0a0 , which conflicts with any installable versions previously reported.
```

**Comment by jdblischak** [comment]
The binaries for [azure-storage-files-datalake-cpp](https://anaconda.org/conda-forge/azure-storage-files-datalake-cpp/files) are being built and uploaded now. If you restart the build in an hour or so, the conda solver error should be resolved.

**Review by h-vetinari — APPROVED** [review]
Thanks for your patience

**Pull request merged**